### PR TITLE
triggers and utils updates

### PR DESCRIPTION
**Pull request by ktbarrett**
More cleanup. I need to probably take a look at the documentation and split up util.py, but that will be a follow on PR or 2. It seemed like a good time to throw up a PR so it didn't get too big.

### CODE REVIEW

File: src/cocotb/triggers.py [hunk 1]
```diff
@@ -27,20 +27,35 @@
 
 """A collections of triggers which a testbench can await."""
```
Comment (cmarqu) [inline src/cocotb/triggers.py:28]:
```suggestion
"""A collection of triggers which a testbench can await."""
```

File: src/cocotb/triggers.py [hunk 2]
```diff
@@ -423,91 +395,90 @@ def __singleton_key__(cls, signal):
...
+    def clear(self) -> None:
         """Clear this event that has fired.
 
         Subsequent calls to :meth:`~cocotb.triggers.Event.wait` will block until
         :meth:`~cocotb.triggers.Event.set` is called again."""
-        self.fired = False
+        self._fired = False
 
     def is_set(self) -> bool:
         """Return true if event has been set"""
```
Comment (cmarqu) [inline src/cocotb/triggers.py:470]:
```suggestion
        """Return ``True`` if event has been set."""
```

File: src/cocotb/triggers.py [hunk 3]
```diff
@@ -598,43 +600,54 @@ class Lock:
...
-        self.locked = False  #: ``True`` if the lock is held.
+    def __init__(self, name: Optional[str] = None) -> None:
+        self._pending_unprimed: List[_Lock] = []
+        self._pending_primed: List[_Lock] = []
+        self.name: Optional[str] = name
+        self._locked: bool = False
+
+    @_locked_back_compat_dec
+    def locked(self) -> bool:
+        """Return True if the lock is locked.
```
Comment (cmarqu) [inline src/cocotb/triggers.py:611]:
```suggestion
        """Return ``True`` if the lock is locked.
```

File: src/cocotb/triggers.py [hunk 4]
```diff
@@ -749,58 +761,62 @@ def __repr__(self):
         return f"{type(self).__qualname__}({self._coroutine!s})"
 
 
-class Waitable(Awaitable):
+class Waitable(Awaitable, Generic[T]):
     """
     Base class for trigger-like objects implemented using coroutines.
 
     This converts a `_wait` abstract method into a suitable `__await__`.
```
Comment (cmarqu) [inline src/cocotb/triggers.py:768]:
```suggestion
    This converts a ``_wait`` abstract method into a suitable ``__await__``.
```

File: src/cocotb/triggers.py [hunk 5]
```diff
@@ -749,58 +761,62 @@ def __repr__(self):
...
     """
 
-    __slots__ = ()
-
-    async def _wait(self):
+    @abstractmethod
+    async def _wait(self) -> T:
         """
         Should be implemented by the sub-class. Called by `await self` to
```
Comment (cmarqu) [inline src/cocotb/triggers.py:774]:
```suggestion
        Should be implemented by the sub-class. Called by ``await self`` to
```

File: src/cocotb/triggers.py [hunk 6]
```diff
@@ -749,58 +761,62 @@ def __repr__(self):
...
 
-class _AggregateWaitable(Waitable):
+_AggregateWaitableSupportedTrigger = Union[
+    Trigger, Waitable, cocotb.task.Task
+]  # TODO: Python 3.8 Add TypeAlias
+
+
+class _AggregateWaitable(Waitable[T]):
     """
     Base class for Waitables that take mutiple triggers in their constructor
```
Comment (cmarqu) [inline src/cocotb/triggers.py:789]:
```suggestion
    Base class for Waitables that take multiple triggers in their constructor.
```

File: src/cocotb/triggers.py [hunk 7]
```diff
@@ -901,10 +915,10 @@ def on_done(ret):
         return await first_trigger  # the first of multiple triggers that fired
 
 
-class ClockCycles(Waitable):
+class ClockCycles(Waitable["ClockCycles"]):
     """Fires after *num_cycles* transitions of *signal* from ``0`` to ``1``."""
```
Comment (cmarqu) [inline src/cocotb/triggers.py:919]:
```suggestion
    """Fire after *num_cycles* transitions of *signal*.
    
    If *rising* is ``True``, count transitions from ``0`` to ``1``;
    if *rising* is ``False``, count transitions from ``1`` to ``0``."""
```

File: src/cocotb/triggers.py [hunk 8]
```diff
@@ -935,12 +949,42 @@ def __repr__(self):
...
 async def with_timeout(
-    trigger: Union[Trigger, Waitable, "cocotb.task.Task", Coroutine[Any, Any, T]],
+    trigger: Union[Trigger, Waitable, cocotb.task.Task, Coroutine],
     timeout_time: Union[float, Decimal],
     timeout_unit: str = "step",
     round_mode: Optional[str] = None,
-) -> T:
+) -> Any:
     r"""
     Waits on triggers or coroutines, throws an exception if it waits longer than the given time.
```
Comment (cmarqu) [inline src/cocotb/triggers.py:989]:
```suggestion
    Wait on triggers or coroutines, throw an exception if it waits longer than the given time.
```

File: src/cocotb/utils.py [hunk 9]
```diff
@@ -78,7 +91,19 @@ def get_sim_time(units: str = "step") -> int:
...
+@overload
+def _ldexp10(frac: Union[float, Fraction], exp: int) -> float: ...
+
+
+@overload
+def _ldexp10(frac: Decimal, exp: int) -> Decimal: ...
+
+
+def _ldexp10(frac: Union[float, Fraction, Decimal], exp: int) -> Union[float, Decimal]:
     """Like math.ldexp, but base 10"""
```
Comment (cmarqu) [inline src/cocotb/utils.py:107]:
```suggestion
    """Like ``math.ldexp``, but using base 10."""
```

File: src/cocotb/utils.py [hunk 10]
```diff
@@ -104,7 +129,10 @@ def get_time_from_sim_steps(steps: int, units: str) -> int:
 
 
 def get_sim_steps(
-    time: Union[Real, Decimal], units: str = "step", *, round_mode: str = "error"
+    time: Union[float, Fraction, Decimal],
+    units: str = "step",
+    *,
+    round_mode: str = "error",
 ) -> int:
     """Calculates the number of simulation time steps for a given amount of *time*.
```
Comment (cmarqu) [inline src/cocotb/utils.py:137]:
```suggestion
    """Calculate the number of simulation time steps for a given amount of *time*.
```

File: src/cocotb/utils.py [hunk 11]
```diff
@@ -159,7 +187,8 @@ def get_sim_steps(
     return result_rounded
 
 
-def _get_log_time_scale(units):
+@lru_cache(maxsize=None)
+def _get_log_time_scale(units: str) -> int:
     """Retrieves the ``log10()`` of the scale factor for a given time unit.
```
Comment (cmarqu) [inline src/cocotb/utils.py:192]:
```suggestion
    """Retrieve the ``log10()`` of the scale factor for a given time unit.
```

**Review by cmarqu — COMMENTED** [review]
More suggestions for docs outside of your actual changes, sorry.
Can't easily make a separate PR from where I am right now (bed, GH GUI :))

**Review by marlonjames — APPROVED** [review]
Everything looks fine to me.
Do you think it would make more sense to add `_LockBackCompat` stuff to 1.9 and do the full switch in 2.0?

**Comment by ktbarrett** [comment]
> Do you think it would make more sense to add _LockBackCompat stuff to 1.9 and do the full switch in 2.0?

Yeah probably... I'll just make the change now and put up a PR on the stable/1.9 branch.

I also need to do some newsfrags, I think.

**Comment by ktbarrett** [comment]
@marlonjames Which changes should be put on 1.9? Just the `.locked` change? Probably also the `.fired` deprecation and also the making private of `_cbhdl`, `_primed`, and `_sim_steps`?

Seems like every change on master should also get a corresponding change on 1.9, giving us twice the work.

File: docs/source/newsfragments/3851.removal.rst [hunk 12]
```diff
@@ -0,0 +1 @@
+Made ``cocotb.triggers.Trigger.primed``, ``cocotb.triggers.GPITrigger.cbhdl``, and ``cocotb.triggers.Timer.sim_steps`` private.``
```
Comment (cmarqu) [inline docs/source/newsfragments/3851.removal.rst:1]:
```suggestion
Made ``cocotb.triggers.Trigger.primed``, ``cocotb.triggers.GPITrigger.cbhdl``, and ``cocotb.triggers.Timer.sim_steps`` private.
```

File: src/cocotb/triggers.py [hunk 13]
```diff
@@ -116,133 +121,124 @@ def _unprime(self):
...
 
-        .. versionchanged:: 1.6
-            Support rounding modes.
+    .. versionchanged:: 1.6
+        Support rounding modes.
 
-        .. versionchanged:: 2.0
-            Passing ``None`` as the *units* argument was removed, use ``'step'`` instead.
+    .. versionchanged:: 2.0
+        Passing ``None`` as the *units* argument was removed, use ``'step'`` instead.
```
Comment (cmarqu) [inline src/cocotb/triggers.py:225]: Sphinx 7.3 (which our RTD build uses) has https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-versionremoved

File: src/cocotb/triggers.py [hunk 14]
```diff
@@ -701,137 +735,125 @@ async def coro_inner():
         result = await Join(task)
         assert result == "Hello world"
 
-    If the coroutine threw an exception, the :keyword:`await` will re-raise it.
-
+    .. note::
+        There is typically no reason to directly use this trigger.
```
Comment (cmarqu) [inline src/cocotb/triggers.py:739]:
```suggestion
        Typically there is no reason to directly use this trigger.
```
(Use wording of `retval` attribute below.)

File: src/cocotb/triggers.py [hunk 15]
```diff
@@ -701,137 +735,125 @@ async def coro_inner():
...
 
-async def _wait_callback(trigger, callback):
-    """
-    Wait for a trigger, and call `callback` with the outcome of the await.
-    """
+async def _wait_callback(
+    trigger: Union[Trigger, Waitable[T], cocotb.task.Task[T]],
+    callback: Callable[[Outcome[T]], None],
+) -> None:
+    """Wait for a trigger, and call `callback` with the outcome of the await."""
```
Comment (cmarqu) [inline src/cocotb/triggers.py:822]:
```suggestion
    """Wait for *trigger*, and call *callback* with the outcome of the await."""
```

**Review by cmarqu — APPROVED** [review]
Sorry for taking so long...

**Pull request merged**